### PR TITLE
[DVT-672] Write to datastore

### PR DESCRIPTION
**Pull request by minhd-vu**
# Description

- Writes blocks, transactions, and events to GCP datastore
- Adds database abstraction interface if writing to another database in the future
- Splits `crawl` and `sensor` functionality to make it clearer
- Introduces `max-peers` flag
- Improves logging

## Jira / Linear Tickets

- [DVT-672](https://polygon.atlassian.net/browse/DVT-672)

# Testing

- [x] Deployed on VMs and observed usage. I did a lot of testing of different VM sizes and the smallest I tested with was a e2-small 0.5-2vCPU and 2GB memory. It ran reliably with around 70% CPU usage and hovered around 40% memory usage (~400MB).

Here is the CPU and memory usage of the VMs. `sensor-1` was running in `eu-west-1` on a 0.5-2vCPU and 2GB memory. `sensor-2` is running in `asia-east-1` on 2CPU and 8GB memory:
Screenshot 2023-05-15 at 9 59 06 AM
Screenshot 2023-05-15 at 9 58 57 AM

Here is how fast the size of datastore is growing:
image
Screenshot 2023-05-12 at 10 17 27 AM
Screenshot 2023-05-12 at 11 13 30 PM
Screenshot 2023-05-15 at 9 57 10 AM

- [x] Tested `crawl`
- [x] Tested `ping`
 
[DVT-672]: https://polygon.atlassian.net/browse/DVT-672?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/p2p/crawl/crawl.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
+/*
+Copyright © 2022 Polygon <engineering@polygon.technology>
```
Comment (rebelArtists) [inline cmd/p2p/crawl/crawl.go:2]: nit date and do we need this in multiple files if repo is public?

File: cmd/p2p/crawl/crawl.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		NodesFile string
+		Database  string
+	}
+)
+
+var (
+	inputCrawlParams crawlParams
+)
+
+// crawlCmd represents the crawl command
```
Comment (rebelArtists) [inline cmd/p2p/crawl/crawl.go:47]: nit comment doesn't lend much detail

File: cmd/p2p/crawl/crawl.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		bn, err := p2p.ParseBootnodes(inputCrawlParams.Bootnodes)
+		if err != nil {
+			log.Error().Err(err).Msg("Unable to parse bootnodes")
+			return err
+		}
+		cfg.Bootnodes = bn
+
+		db, err := enode.OpenDB(inputCrawlParams.Database)
+		if err != nil {
+			return err
```
Comment (rebelArtists) [inline cmd/p2p/crawl/crawl.go:73]: may want to standardize log.Error() as we did with earlier condition before return

File: cmd/p2p/sensor/sensor.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+		bn, err := p2p.ParseBootnodes(inputSensorParams.Bootnodes)
+		if err != nil {
+			log.Error().Err(err).Msg("Unable to parse bootnodes")
+			return err
+		}
+		cfg.Bootnodes = bn
+
+		db, err := enode.OpenDB(inputSensorParams.Database)
+		if err != nil {
+			return err
```
Comment (rebelArtists) [inline cmd/p2p/sensor/sensor.go:94]: same suggestion to standardize log.Error() for all conditionals

File: cmd/p2p/sensor/sensor.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	if err := SensorCmd.MarkPersistentFlagRequired("bootnodes"); err != nil {
+		log.Error().Err(err).Msg("Failed to mark bootnodes as required persistent flag")
+	}
+	SensorCmd.PersistentFlags().IntVarP(&inputSensorParams.Threads, "parallel", "p", 16, "How many parallel discoveries to attempt.")
+	SensorCmd.PersistentFlags().IntVarP(&inputSensorParams.NetworkID, "network-id", "n", 0, "Filter discovered nodes by this network ID.")
+	if err := SensorCmd.MarkPersistentFlagRequired("network-id"); err != nil {
+		log.Error().Err(err).Msg("Failed to mark network-id as required persistent flag")
+	}
+	SensorCmd.PersistentFlags().StringVarP(&inputSensorParams.Database, "database", "d", "", "Node database for updating and storing client information.")
+	SensorCmd.PersistentFlags().StringVarP(&inputSensorParams.ProjectID, "project-id", "P", "", "GCP project ID.")
```
Comment (rebelArtists) [inline cmd/p2p/sensor/sensor.go:130]: what if someone's using aws?

File: p2p/rlpx.go [hunk 6]
```diff
@@ -17,17 +17,22 @@
...
+	"container/list"
 	"fmt"
+	"math/rand"
 	"net"
 	"strings"
+	"sync/atomic"
 	"time"
 
 	"github.com/ethereum/go-ethereum/common"
+	// "github.com/ethereum/go-ethereum/core/types"
```
Comment (rebelArtists) [inline p2p/rlpx.go:29]: nit drop

**Review by rebelArtists — COMMENTED** [review]
overall really nice code separation and clarity, left just a couple nits and suggestions

File: p2p/p2p.go [hunk 7]
```diff
@@ -5,13 +5,37 @@ import (
...
 
 	"github.com/ethereum/go-ethereum/p2p/enode"
 	"github.com/ethereum/go-ethereum/p2p/enr"
 	"github.com/ethereum/go-ethereum/rlp"
 )
 
+func Listen(ln *enode.LocalNode) (*net.UDPConn, error) {
+	addr := "0.0.0.0:0"
+
+	socket, err := net.ListenPacket("udp4", addr)
```
Comment (praetoriansentry) [inline p2p/p2p.go:19]: Hmm. I'm not familiar with using `:0` here. What would that do?

Comment (praetoriansentry) [reply p2p/p2p.go:19]: Tactically, this might need to be configurable

File: p2p/rlpx.go [hunk 8]
```diff
@@ -151,97 +156,236 @@ loop:
...
 				res := &BlockHeaders{
 					RequestId: msg.RequestId,
 				}
 				if err := c.Write(res); err != nil {
 					c.logger.Error().Err(err).Msg("Failed to write BlockHeaders response")
+					return err
 				}
 			case *BlockBodies:
-				c.logger.Info().Msgf("Received %v block bodies", len(msg.BlockBodiesPacket))
+				atomic.AddInt32(&count.BlockBodies, int32(len(msg.BlockBodiesPacket)))
```
Comment (praetoriansentry) [inline p2p/rlpx.go:220]: These counters feel like they should be openmetrics / prometheus logs.. Maybe a future improvement.

File: p2p/rlpx.go [hunk 9]
```diff
@@ -151,97 +156,236 @@ loop:
...
+	// contain information about the block hash.
+	requests := list.New()
+	var requestNum uint64 = 0
+
+	// dbCh is used to limit the number of database goroutines running at one
+	// time with a buffered channel. Without this, a large influx of messages can
+	// bog down the system and leak memory.
+	var dbCh chan struct{}
+	if db != nil {
+		dbCh = make(chan struct{}, db.MaxConcurrentWrites())
```
Comment (praetoriansentry) [inline p2p/rlpx.go:178]: nice

**Review by praetoriansentry — APPROVED** [review]

File: p2p/database/datastore.go [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+	key := datastore.NameKey("blocks", block.Hash().Hex(), nil)
+	var dsBlock DatastoreBlock
+	_ = d.client.Get(d.ctx, key, &dsBlock)
+
+	if dsBlock.DatastoreHeader == nil {
+		dsBlock.DatastoreHeader = newDatastoreHeader(block.Header())
+	}
+
+	if dsBlock.Transactions == nil && len(block.Transactions()) > 0 {
+		d.writeTransactions(block.Transactions())
```
[the remark this answers is not in the document]
Comment (minhd-vu) [reply p2p/database/datastore.go:114]: Great catch!

File: p2p/database/datastore.go [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+	if err == nil {
+		from = msg.From().Hex()
+	}
+
+	if tx.To() != nil {
+		to = tx.To().Hex()
+	}
+
+	return &DatastoreTransaction{
+		Data:      tx.Data(),
```
[the remark this answers is not in the document]
Comment (minhd-vu) [reply p2p/database/datastore.go:246]: The transaction hash is used as the key. It's not directly stored in the struct but used when:
```go
keys = append(keys, datastore.NameKey("transactions", tx.Hash().Hex(), nil))
```
Here's datastore's view with the hash parts highlighted. You can also query from this key and I have done it before. This structure is used again with the `blocks` entities as well.
Screenshot 2023-05-16 at 12 20 22 PM

File: p2p/database/datastore.go [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+}
+
+// DatastoreEvent can represent a peer sending the sensor a transaction hash or
+// a block hash.
+type DatastoreEvent struct {
+	SensorId string
+	PeerId   string
+	Hash     *datastore.Key
+	Time     time.Time
+}
```
[the remark this answers is not in the document]
Comment (minhd-vu) [reply p2p/database/datastore.go:33]: I don't think so. These events are being written to different tables, but I understand why that may not be clear. I'll add some more comments to make this clearer. Here are some screenshots:

Screenshot 2023-05-16 at 12 33 28 PM
Screenshot 2023-05-16 at 12 32 59 PM

File: cmd/p2p/p2p.go [hunk 13]
```diff
@@ -1,18 +1,47 @@
...
+	} else if verbosity < 400 {
+		zerolog.SetGlobalLevel(zerolog.WarnLevel)
+	} else if verbosity < 500 {
+		zerolog.SetGlobalLevel(zerolog.InfoLevel)
+	} else if verbosity < 600 {
+		zerolog.SetGlobalLevel(zerolog.DebugLevel)
+	} else {
+		zerolog.SetGlobalLevel(zerolog.TraceLevel)
+	}
+}
```
[the remark this answers is not in the document]
Comment (minhd-vu) [reply cmd/p2p/p2p.go:47]: I was thinking about doing a persistent flag in the `rootCmd`, but I noticed that `loadtest` had a different default. Maybe we should standardize the default log level in another ticket.

**Pull request merged**